retry.go

```diff
@@ -287,4 +287,55 @@ func (t *Transaction[T]) Process(state T) (T, error) {
 	return state, err
 }
 
-// throttle ?
+type throttle struct {
+	mu         *sync.Mutex
+	timer      *time.Timer
+	needInvoke bool
+	interval   time.Duration
+	callbacks  []func()
+}
+
+func (th *throttle) throttledFunc() {
+	th.mu.Lock()
+	defer th.mu.Unlock()
+	th.needInvoke = true
+	if th.timer == nil {
+		th.timer = time.AfterFunc(th.interval, func() {
+			th.purge(false)
+		})
+	}
+}
+
+func (th *throttle) purge(forcePurge bool) {
+	th.mu.Lock()
+	defer th.mu.Unlock()
+
+	if th.timer != nil {
+		th.timer.Stop()
+	}
+
+	if th.needInvoke || forcePurge {
+		for _, f := range th.callbacks {
+			go f()
```

Comment:
by default, i would make it synchronous and let the developer trigger a goroutine

WDYT ?

Reply:
can you explain why you trigger the callbacks when calling purge() ?

IMO, to force a call, we would code something like this:

```go
throttle()
throttle() // skipped
throttle() // skipped
reset()
throttle()
```

Reply:
> by default, i would make it synchronous and let the developer trigger a goroutine
> 
> WDYT ?

Your way is more simple and configuable, 
I'll fix it that way.


> can you explain why you trigger the callbacks when calling purge() ?
> 
> IMO, to force a call, we would code something like this:
> 
> ```go
> throttle()
> throttle() // skipped
> throttle() // skipped
> reset()
> throttle()
> ```

I wanted to offer a function to force triggering callbacks and reset current interval schedule.
The usage of it is exactly same as you mentioned.
I thought, name 'reset' might imply resetting interval schedule, but not including calling the functions